File: 0117. 填充每个节点的下一个右侧节点指针 II/BFS.js
// 112ms
var connect = function (root) {
    if (root != null) {
        let q = [[root, 0]], pre = [root, -1]
        while (q.length != 0) {
            let [node, level] = q.shift()
            pre[0].next = level == pre[1] ? node : null
            pre = [node, level]
            if (node.left != null) {
                q.push([node.left, level + 1])
            }
            if (node.right != null) {
                q.push([node.right, level + 1])
            }
        }
    }
    return root
};

File: 0116. 填充每个节点的下一个右侧节点指针/BFS.swift
// 108ms
class Solution {
    func connect(_ root: Node?) -> Node? {
        var pre: Node?, preLevel = -1
        var queue = [(Node, Int)]()
        if let root = root {
            queue.append( (root, 0) )
            while !queue.isEmpty {
                let (node, level) = queue.removeFirst()
                pre?.next = level == preLevel ? node : nil
                pre = node
                preLevel = level
                if let left = node.left {
                    queue.append( (left, level + 1) )
                }
                if let right = node.right {
                    queue.append( (right, level + 1) )
                }
            }
        }
        return root
    }
}